### PR TITLE
fix: Validate groups in convolution meta kernel

### LINKED ISSUE #178472: conv1d on meta device path skips groups validation and returns an invalid output shape
## Summary
`torch.nn.functional.conv1d` on the `meta` device does not validate `groups`/weight compatibility the same way as the normal backend. For an invalid configuration, the meta path returns a tensor shape instead of raising an error.

## Reproduction
```python
import torch
import torch.nn.functional as F

x = torch.randn(2, 12, 30, device='meta')
w = torch.randn(3, 2, 3, device='meta')
print(F.conv1d(x, w, groups=6).shape)
```

## Expected behavior
The call should raise a validation error, matching the regular backend, because `groups=6` is incompatible with `weight.shape[0]=3` / input channels.

## Actual behavior
The meta backend returns:

```text
torch.Size([2, 3, 28])
```

without reporting the invalid `groups` configuration.

### Versions

Collecting environment information...
PyTorch version: 2.10.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.11-060511-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000

Nvidia driver version: 580.65.06
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             112
On-line CPU(s) list:                0-111
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz
CPU family:                         6
Model:                              106
Thread(s) per core:                 2
Core(s) per socket:                 28
Socket(s):                          2
Stepping:                           6
CPU(s) scaling MHz:                 27%
CPU max MHz:                        3100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4000.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          2.6 MiB (56 instances)
L1i cache:                          1.8 MiB (56 instances)
L2 cache:                           70 MiB (56 instances)
L3 cache:                           84 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95,97,99,101,103,105,107,109,111
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.1
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] optree==0.18.0
[pip3] torch==2.10.0
[pip3] triton==3.6.0
[conda] numpy                     2.4.1                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] optree                    0.18.0                   pypi_0    pypi
[conda] torch                     2.10.0                   pypi_0    pypi
[conda] triton                    3.6.0                    pypi_0    pypi

cc @malfet @ezyang @eellison @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by harshaljanjani**
### What does this PR do?

→ The bug is in [calc_conv_nd_return_shape](https://github.com/pytorch/pytorch/blob/5339e63cc08f87eb4ba1fe151164f0bb5e429c51/torch/_meta_registrations.py#L2682-L2686); the non-transposed branch only checked input channels and omitted the checks the real backend enforces, so an invalid config silently gave `torch.Size(foo)` instead of raising.
→ Fixed by adding the missing checks ahead of the channel check mirroring the C++ [check_shape_forward](https://github.com/pytorch/pytorch/blob/5339e63cc08f87eb4ba1fe151164f0bb5e429c51/aten/src/ATen/native/Convolution.cpp#L687-L700) so the meta error is now identical to the eager backend :)

Fixes #178472. This also fixes a related `groups=0` bug in the same meta helper. It returned a transposed shape and raised `Invalid channel dimensions` instead of matching eager's `"non-positive groups is not supported"` error. Builds on the closed/stale #178563; credited the original author as a co-author.

cc @mikaylagawarecki @ezyang @albanD

### Verification and Testing

→ <ins>Locally</ins> full `-k conv` meta suite (269 tests) and the new `TestMetaKernelConv` test pass with no new failures.

```
python test/test_meta.py -k "TestMetaKernelConv" -v
python test/test_meta.py -k "conv" -v
```

**Before (meta silently returns a shape)**

<br>

**After (matches the eager backend exactly)**

### Environment

* `torch` version: `2.14.0a0+git5339e63` (source build, `USE_CUDA=0`)
* Platform: `Linux-6.8.0-1064-gcp-x86_64-with-glibc2.35`
* Python version: `3.10.12`
* The bug and fix are device-independent (raised from the `@register_meta` shape function); reproduced and fixed on CPU.

### CODE REVIEW

**Comment by harshaljanjani** [comment]
@jbschlosser Good day! Could you please triage this issue at your convenience? Thank you :)

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!

**Comment by harshaljanjani** [comment]
I understand; thanks for letting me know! I'll be on the lookout for the label the next time around :)
